packages/network-controller/src/NetworkController.ts

```diff
@@ -481,6 +539,131 @@ export class NetworkController extends BaseControllerV2<
       });
     }
   }
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param networkConfiguration - The network configuration to add or, if rpcUrl matches an existing entry, to modify.
+   * @param networkConfiguration.rpcUrl -  RPC provider url.
+   * @param networkConfiguration.chainId - Network ID as per EIP-155.
+   * @param networkConfiguration.ticker - Currency ticker.
+   * @param networkConfiguration.nickname - Personalized network name.
+   * @param networkConfiguration.rpcPrefs - Personalized preferences (i.e. preferred blockExplorer)
+   * @param options - additional configuration options.
+   * @param options.setActive - An option to set the newly added networkConfiguration as the active provider.
+   * @param options.referrer - The site from which the call originated, or 'metamask' for internal calls - used for event metrics.
+   * @param options.source - Where the upsertNetwork event originated (i.e. from a dapp or from the network form)- used for event metrics.
+   * @returns id for the added or updated network configuration
+   */
+  upsertNetworkConfiguration(
+    { rpcUrl, chainId, ticker, nickname, rpcPrefs }: NetworkConfiguration,
+    {
+      setActive = false,
+      referrer,
+      source,
+    }: { setActive?: boolean; referrer: string; source: string },
+  ) {
+    assertIsStrictHexString(chainId);
+
+    if (!isSafeChainId(parseInt(chainId, 16))) {
+      throw new Error(
+        `Invalid chain ID "${chainId}": numerical value greater than max safe value.`,
```

Comment:
Nit: It looks like this validation isn't covered by unit tests, nor is the metric event. It would be great to include complete coverage of these new functions.

Reply:
added all of the `upsertNetworkConfiguration` tests from the extension side: https://github.com/MetaMask/core/pull/1064/commits/0124a625a57d81ae00c1dcaa03e880c62f6ff14e#diff-3439880e0b961116f2219977efca0fe970b9fa919ac62b5ad934c2274edf206dR3588